### PR TITLE
[browser][bindings] Fix error with SharedArrayBuffer when used as a backing view.

### LINKED ISSUE #46624: [browser][bindings] "Object '...' is not a typed array" in js_typed_array_to_array.
```C#
public static int TestSum(SharedArrayBuffer sharedArrayBuffer)
{
    Int32Array array = new Int32Array(sharedArrayBuffer);
    Span<int> nativeArray = array; // error                        
    
    int sum = 0;
    for (int i = 0; i < nativeArray.Length; i++)
    {
        sum += nativeArray[i];
    }

    return sum;
}
```

Fails with "Object '...' is not a typed array" in js_typed_array_to_array. 

Originally reported here: https://github.com/dotnet/runtime/issues/46016

This can reliably be reproduced and changes will be coming soon.

### PR DESCRIPTION
**Pull request by kjpou1**
Fails with "Object '...' is not a typed array" when used as a backing view for TypedArrays.

```C#
public static int TestSum(SharedArrayBuffer sharedArrayBuffer)
{
    Int32Array array = new Int32Array(sharedArrayBuffer);
    Span<int> nativeArray = array; // error                        
    
    int sum = 0;
    for (int i = 0; i < nativeArray.Length; i++)
    {
        sum += nativeArray[i];
    }

    return sum;
}
```

Fixes https://github.com/dotnet/runtime/issues/46624

Thanks @benaadams for the firefox support !!!

### CODE REVIEW

File: src/mono/wasm/runtime/binding_support.js [hunk 1]
```diff
@@ -497,7 +497,7 @@ var BindingSupportLib = {
 			// you need to use a view. A view provides a context — that is, a data type, starting offset, 
 			// and number of elements — that turns the data into an actual typed array.
 			// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays
-			if (!!(js_obj.buffer instanceof ArrayBuffer && js_obj.BYTES_PER_ELEMENT)) 
+			if (!!((js_obj.buffer instanceof ArrayBuffer || js_obj.buffer instanceof SharedArrayBuffer) && js_obj.BYTES_PER_ELEMENT))
```
Comment (benaadams) [inline src/mono/wasm/runtime/binding_support.js:500]: Needs more guarding around the type? e.g. will throw in FireFox
```
Uncaught ReferenceError: SharedArrayBuffer is not defined
```

Comment (kjpou1) [reply src/mono/wasm/runtime/binding_support.js:500]: 👍

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/SharedArrayBuffer.cs [hunk 2]
```diff
@@ -23,6 +23,22 @@ internal SharedArrayBuffer(IntPtr jsHandle, bool ownsHandle) : base(jsHandle, ow
         /// <value>The size, in bytes, of the array.</value>
         public int ByteLength => (int)GetObjectProperty("byteLength");
 
+        /// <summary>
+        /// Returns a new JavaScript Core SharedArrayBuffer whose contents are a copy of this SharedArrayBuffer's bytes from begin = 0,
+        /// inclusive, through to the end of the sequence, exclusive.
+        /// </summary>
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/SharedArrayBuffer.cs:29]: This method doesn't take a `begin`, or `end` args, so this description needs to be reworded.

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj [hunk 3]
```diff
@@ -8,14 +8,15 @@
     <Compile Include="System\Runtime\InteropServices\JavaScript\JavaScriptTests.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\DataViewTests.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\TypedArrayTests.cs" />
+    <Compile Include="System\Runtime\InteropServices\JavaScript\SharedArrayBufferTests.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\ArrayTests.cs" />
-    <!-- <Compile Include="System\Runtime\InteropServices\JavaScript\MapTests.cs" /> -->
+    <Compile Include="System\Runtime\InteropServices\JavaScript\MapTests.cs" />
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj:13]: Why is this changed needed here?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj [hunk 4]
```diff
@@ -8,14 +8,15 @@
...
     <Compile Include="System\Runtime\InteropServices\JavaScript\MarshalTests.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\HelperMarshal.cs" />
-    <Compile Include="System\Runtime\InteropServices\JavaScript\Http\HttpRequestMessageTest.cs" />    
+    <Compile Include="System\Runtime\InteropServices\JavaScript\Http\HttpRequestMessageTest.cs" />
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj:16]: Why is this changed needed here?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices.JavaScript;
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:5]: Unused

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+        public static IEnumerable<object[]> Object_Prototype()
+        {
+            _objectPrototype ??= new Function("return Object.prototype.toString;");
+            yield return new object[] { _objectPrototype.Call() };
+        }
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
+        public static void SharedArrayBufferLength(Function objectPrototype)
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:22]: Maybe `SharedArrayBuffer_ZeroLength`?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(0);
+            Assert.Equal("[object SharedArrayBuffer]", objectPrototype.Call(d));
+            Assert.Equal(0, d.Length);
+        }
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
+        public static void SharedArrayBufferLength1(Function objectPrototype)
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:31]: Maybe `SharedArrayBuffer_NonZeroLength`?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            Assert.Equal("[object SharedArrayBuffer]", objectPrototype.Call(d));
+            Assert.Equal(0, d.Length);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferLength2()
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal(50, d.ByteLength);
+        }
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:43]: Dupe of `SharedArrayBufferLength1`?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+        [Fact]
+        public static void SharedArrayBufferSlice()
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal(50, d.Slice().ByteLength);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferSlice1()
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:53]: Maybe `SharedArrayBuffer_Slice_BeginEndForFullArray`, or something better?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal(50, d.Slice(0, 50).ByteLength);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferSlice2()
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:60]: Maybe `SharedArrayBuffer_Slice_BeginZero`?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal(50, d.Slice(0).ByteLength);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferSlice3()
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:67]: Maybe SharedArrayBuffer_Slice_BeginNegative?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal(3, d.Slice(-3).ByteLength);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferSlice4()
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:74]: Maybe SharedArrayBuffer_Slice_BeginEndSubset?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            Int32Array sliced = new Int32Array(buffer.Slice(4,12));
+            // expected output: Int32Array [42, 0]
+
+            Assert.Equal(2, sliced.Length);
+            Assert.Equal(42, sliced[0]);
+            Assert.Equal(0, sliced[1]);
+        }
+
+        [Fact]
+        public static void SharedArrayBufferSliceAndDice1()
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:102]: Maybe `SharedArrayBufferSliceAndDiceAndUseThroughSpan`?

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+            int32View[1] = 42;
+
+            Assert.Equal(4, int32View.Length);
+            Assert.Equal(42, int32View[1]);
+
+            Int32Array sliced = new Int32Array(buffer.Slice(4,12));
+            // expected output: Int32Array [42, 0]
+
+            Span<int> nativeArray = sliced; // error
```
Comment (radical) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:117]: What do you mean by `error` here?

**Review by radical — CHANGES_REQUESTED** [review]

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            yield return new object[] { _objectPrototype.Call() };
+        }
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
+        public static void SharedArrayBufferLength(Function objectPrototype)
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(0);
+            Assert.Equal("[object SharedArrayBuffer]", objectPrototype.Call(d));
+            Assert.Equal(0, d.Length);
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:26]: No this is not correct.  Length will always return 0 but it is ByteLength that is actually defined.  I can take out this test if it is causing problems.

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
+        public static void SharedArrayBufferLength1(Function objectPrototype)
+        {
+            SharedArrayBuffer d = new SharedArrayBuffer(50);
+            Assert.Equal("[object SharedArrayBuffer]", objectPrototype.Call(d));
+            Assert.Equal(0, d.Length);
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:35]: No this is not correct. `Length` will always return 0 but it is `ByteLength` that is actually defined.

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (radical) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:26]: So, `.Length` is not defined here, right? So, if it returns `undefined`, then what are we really testing here?

**Comment by radical** [comment]
These might be useful:
```csharp
        private static TheoryData<SharedArrayBuffer> GetTestData(int length)
        {
            // create a SharedArrayBuffer with a size in bytes
            SharedArrayBuffer buffer = new SharedArrayBuffer(length);
            Int32Array int32View = new Int32Array(buffer);  // create view
            for (int i = 0; i < int32View.Length; i ++)
                int32View[i] = i + 10;

            int32View[1] = 42;
            return new TheoryData<SharedArrayBuffer> { buffer };
        }

        [Theory]
        [MemberData(nameof(GetTestData), 16)]
        public static void SharedArrayBufferSliceAndDice3_Subset(SharedArrayBuffer buffer)
        {
            Int32Array sliced = new Int32Array(buffer.Slice(4,12));

            Assert.Equal(2, sliced.Length);
            Assert.Equal(42, sliced[0]);
            Assert.Equal(12, sliced[1]);
        }

        [Theory]
        [MemberData(nameof(GetTestData), 16)]
        public static void SharedArrayBufferSliceAndDice3_SubsetFromTheBack(SharedArrayBuffer buffer)
        {
            Int32Array sliced = new Int32Array(buffer.Slice(-4));

            Assert.Equal(1, sliced.Length);
            Assert.Equal(13, sliced[0]);
        }

        [Theory]
        [MemberData(nameof(GetTestData), 16)]
        public static void SharedArrayBufferSliceAndDice3_SubsetFromTheBackWithEnd(SharedArrayBuffer buffer)
        {
            Int32Array sliced = new Int32Array(buffer.Slice(-12, -4));

            Assert.Equal(2, sliced.Length);
            Assert.Equal(42, sliced[0]);
            Assert.Equal(12, sliced[1]);
        }
```

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 6 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:22]: Removed

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs:26]: Not a problem it is removed.

**Comment by kjpou1** [comment]
Thanks Ankit.  Will incorporate

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 5 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 14 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 13 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 12 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 11 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 10 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 7 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 8 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs (hunk 9 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/SharedArrayBufferTests.cs]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj (hunk 3 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj (hunk 4 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System.Private.Runtime.InteropServices.JavaScript.Tests.csproj]: Addressed in recent commits

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/SharedArrayBuffer.cs (hunk 2 above)
Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/SharedArrayBuffer.cs]: Addressed in recent commits

File: src/mono/wasm/runtime/binding_support.js (hunk 1 above)
Comment (benaadams) [inline src/mono/wasm/runtime/binding_support.js:500]:
```suggestion
			if (!!((js_obj.buffer instanceof ArrayBuffer ||  (typeof SharedArrayBuffer !== "undefined" && js_obj.buffer instanceof SharedArrayBuffer)) && js_obj.BYTES_PER_ELEMENT))
```
This should fix the exception from the `instanceof` I believe

Comment (kjpou1) [reply src/mono/wasm/runtime/binding_support.js]: Thanks @benaadams for the suggested change.  I did it just a little different but please let me know if this works or you have a problem with the way I approached it.

Comment (kjpou1) [reply src/mono/wasm/runtime/binding_support.js]: Please see recent commit that should address this.  Thank you for the heads up.  Totally forgot about the non supported.

**Review by radical — APPROVED** [review]
LGTM! 👍

**Review by benaadams — APPROVED** [review]

**Pull request merged**